File: anemometer.py
import RPi.GPIO as GPIO
from collections import deque
import time
import json

"""
 Based on:
   https://www.john.geek.nz/2011/07/la-crosse-tx20-anemometer-communication-protocol/
"""

#GPIO.setmode(GPIO.BOARD)
#pin = 3

GPIO.setmode(GPIO.BCM)
pin = 2

GPIO.setup(pin, GPIO.IN)

wind_dir_list = {
  0: "N",
  1: "NNE",
  2: "NE",
  3: "ENE",
  4: "E",
  5: "ESE",
  6: "SE",
  7: "SSE",
  8: "S",
  9: "SSW",
  10: "SW",
  11: "WSW",
  12: "W",
  13: "WNW",
  14: "NW",
  15: "NNW"
}

def parse_frame(frame):
  #print frame
  frame_header = frame[0:5]
  #print json.dumps(frame_header)
  if json.dumps(frame_header) <> "[1, 1, 0, 1, 1]":
    #print "invalid frame header"
    return

  print

  wind_dir = frame[5:9]
  wind_dir.reverse()
  wind_speed = frame[9:21]
  wind_speed.reverse()
  wind_dir_minv = [0 if b == 1 else 1 for b in wind_dir]
  wind_speed_minv = [0 if b == 1 else 1 for b in wind_speed]
  checksum = frame[21:25]
  wind_dir_inv = frame[25:29]
  wind_dir_inv.reverse()
  wind_speed_inv = frame[29:41]
  wind_speed_inv.reverse()

  print "dir:   ", wind_dir, wind_dir_minv
  #print "inv dir:   ", wind_dir_inv
  if wind_dir_minv != wind_dir_inv:
    print "wind direction mismatch"
    #return

  print "speed: ", wind_speed, wind_speed_minv
  #print "inv speed: ", wind_speed_inv
  if wind_speed_minv != wind_speed_inv:
    print "wind speed mismatch"
    #return

  # TODO: verify checksum
  #print "checksum: ", checksum

  wind_dir_int = int("".join([str(b) for b in wind_dir_minv]), 2)
  wind_speed_int = int("".join([str(b) for b in wind_speed_minv]), 2)
  print "dir int:   %d => %s" % (wind_dir_int, wind_dir_list[wind_dir_int])
  print "speed int: %d => %.1f meter/sec" % (wind_speed_int, wind_speed_int * 0.1)
  

raw_frame = []
first = 0
freq_delay = 0.00093
while True:
  start = int(round(time.time() * 1000000))
  v = GPIO.input(pin)

  if len(raw_frame) >= 41:
    # Data raw_frame is 41 bits long and last about 49 msec
    if len(raw_frame) > 0:
      #print "Data Frame (length:%d):" % len(raw_frame)
      frame = raw_frame[0:41]
      parse_frame(frame)
    raw_frame = []
    first = 0
  freq = None
  if len(raw_frame) == 0 and v == 1:
    first = int(round(time.time() * 1000000))
    freq = freq_delay + freq_delay/3
  if first > 0:
    raw_frame.append(v)
    freq = freq_delay

  #time.sleep(0.00095)  # not too bad
  #time.sleep(0.00094)  # good too
  #time.sleep(0.00093)  # best ?
  if freq: time.sleep(freq)

File: anemometer.c
#include <stdio.h>
#include <wiringPi.h>
#include <time.h>
#include <sys/time.h>
#include <sched.h>
#include <string.h>
#include <math.h>

/*
 * Requirements:
 *   apt-get install gcc wiringpi
 *
 * Build command:
 *   gcc -o anemometer anemometer.c -lwiringPi -lm
 *
 * Run:
 *   nice -19 ./anemometer
 *
 * Based on:
 *   https://www.john.geek.nz/2011/07/la-crosse-tx20-anemometer-communication-protocol/
 */

long long current_timestamp() {
    struct timeval te; 
    gettimeofday(&te, NULL);
    long long microseconds = te.tv_sec*1000000LL + te.tv_usec;
    return microseconds;
}

int decode(int frame[], int start, int end, int inverted) {
  int i;
  int v = 0;
  for( i = start; i < end; i++ ) {
    //printf("v:%d", frame[i]);
    v = v | (abs(inverted - frame[i]) * (int)pow(2.0, (double)(i - start)));
  }
  return v;
}

void decode_frame(int frame[]) {
      int header, wind_dir, wind_speed, checksum, wind_dir_inv, wind_speed_inv, i;

      for( i=0; i<=41; i++ ) {
        printf(" %d ", frame[i]);
      }
      printf("\n");
      /*
      */

      header = decode(frame, 0, 5, 1);
      if( header != 4 ) {
        printf("Invalid frame: bad header:[%d]\n", header);
        return;
      }

      wind_dir = decode(frame, 5, 9, 1);
      wind_speed = decode(frame, 9, 21, 1);
      checksum = decode(frame, 21, 25, 1);
      wind_dir_inv = decode(frame, 25, 29, 0);
      wind_speed_inv = decode(frame, 29, 41, 0);

      if( wind_dir != wind_dir_inv ) {
        printf("Invalid frame: wind direction mismatch. wind_dir:[%d] wind_dir_inv:[%d]\n", wind_dir, wind_dir_inv);
        return;
      }
      if( wind_speed != wind_speed_inv ) {
        printf("Invalid frame: wind speed mismatch. wind_speed:[%d] wind_speed_inv:[%d]\n", wind_speed, wind_speed_inv);
        return;
      }
      int sum = 0;
      for( i = 5; i < 21; i+=4 ) {
        int v = decode(frame, i, i+4, 1);
        sum += v;
      }
      sum = sum % 16;
      if( sum != checksum ) {
        printf("Invalid frame: bad checksum. sum:[%d] checksum:[%d]\n", sum, checksum);
        return;
      }
      printf("header:[%d] wind_dir:[%d] wind_dir_inv:[%d] wind_speed:[%d] wind_speed_inv:[%d] checksum:[%d]\n", header, wind_dir, wind_dir_inv, wind_speed, wind_speed_inv, checksum);
      fflush(stdout);
}


int main (void) {
  struct timeval tv;
  double t2;
  int i, v;
  struct timespec req;
  req.tv_nsec = 1000;
  req.tv_sec = 0;
  double diff;

  struct sched_param rt_param;
  int rt_max_prio = sched_get_priority_max(SCHED_FIFO);
  sched_getparam(getpid(), &rt_param);
  rt_param.sched_priority = rt_max_prio;
  sched_setscheduler(getpid(), SCHED_FIFO, &rt_param);

  printf("Raspberry Pi anemometer\n");
 
  if (wiringPiSetup () == -1) {
    return 1;
  }
 
  pinMode(8, INPUT);

  double first = 0;
  int pos = 0;
  int frame[50];
  int delay;
  int signal_freq = 1210;
  for (;;) {
    v = digitalRead(8);
    if( first > 0 && ((current_timestamp() - first) > 49300) ) {
      printf("\nNEW DATA FRAME usec:[%d] bits:[%d]:\n", (int)(current_timestamp() - first), pos);
      decode_frame(frame);
      first = 0;
      pos = 0;
    }
    // Waiting for signal
    delay = 120;
    if( first == 0 && v == 1 ) {
      first = current_timestamp();
      // Decoding, first time
      delay = signal_freq + signal_freq/2 - delay;
    }
    if( first > 0 ) {
      frame[pos++] = v;
      // Decoding
      delay = signal_freq;
    }

    diff = current_timestamp() - t2;
    while( diff < delay ) {
        nanosleep(&req, NULL);
        diff = current_timestamp() - t2;
    }
    t2 = current_timestamp();
    //printf("%f :[%d]\n", diff, v);
  }
  return 0;
}
